Extend virtctl to Support Interactive SEV Attestation

Fuzzy-Math opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**: 

This PR extends `virtctl` to support interactive attestation of SEV enabled workloads.

The following subcommands have been added:
- `virtctl sev fetch-cert-chain`
- `virtctl sev setup-session`
- `virtctl sev query-measurement`
- `virtctl sev inject-secret`

**Special notes for your reviewer**:
- This PR relies on #7197 for intial SEV enablement
- Implementation based off initial [design proposal](https://github.com/kubevirt/community/blob/main/design-proposals/amd-sev.md)
- Until #7197 is merged, another [branch](https://github.com/Fuzzy-Math/kubevirt/tree/sev-base) is being maintained to allow building and testing of these changes

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Extending virtctl to support interactive attestation of AMD SEV enabled workloads
```

Fuzzy-Math commented:
/hold

vasiliy-ul commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,290 @@
+			if err != nil {
+				return fmt.Errorf("Cannot marshal SEV platform info %v", err)
+			}
+			if pathProvided {
+				if _, err := file.Write(data); err != nil {
+					log.Log.Errorf("[SEV] Failed to write certificate chain to %s", outpath)
+					return fmt.Errorf("Error %s/%s: %v", namespace, vmiName, err)
+				}
+				log.Log.Infof("[SEV]Writing platform certificate chain to %s", outpath)
+			} else {
+				fmt.Printf("%s\n", string(data))
+			}
```
I think the code can be simplified by using https://pkg.go.dev/io/ioutil#WriteFile . That way you will not need the `file, err = os.OpenFile(...)` part above. Same applies for the `COMMAND_QUERY_MEASUREMENT` case.

Fuzzy-Math replied:
Using `WriteFile()` would overwrite the file if it already existed. As designed, I don't allow overwriting existing files and it's a failure condition if the file exists.

vasiliy-ul commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,289 @@
+		vm = kubecli.NewMinimalVM(vmiName)
+		vmi = api.NewMinimalVMI(vmiName)
+		ctrl = gomock.NewController(GinkgoT())
+		workDir, err = ioutil.TempDir("", "sev-test")
```
You can use `GinkgoT().TempDir()` here and you will not need to do the cleanup in `AfterEach`

alicefr replied:
Not sure why this was resolved but I'd also suggest to used the TempDIr creation from ginkgo or any particular reason for not using it

vasiliy-ul commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,289 @@
+		Expect(err).To(BeNil())
+		kubecli.GetKubevirtClientFromClientConfig = kubecli.GetMockKubevirtClientFromClientConfig
+		kubecli.MockKubevirtClientInstance = kubecli.NewMockKubevirtClient(ctrl)
+		vmInterface = kubecli.NewMockVirtualMachineInterface(ctrl)
```
This seems to be unused in the tests. Maybe its not needed?

vasiliy-ul commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,289 @@
+
+				cmd := tests.NewVirtctlCommand(sev.COMMAND_SEV, sev.COMMAND_FETCH_CERT_CHAIN, vmiName, "--output", fpath)
+				Expect(cmd.Execute()).To(BeNil())
+				// Expect file creation
```
what about also checking that the content of the file matches `sevPlatformInfo` returned by the mock? Same for the measurement call.

vasiliy-ul reviewed (commented):
Looks good. Just added a few minor comments.

vasiliy-ul commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,280 @@
+			}
+			if outpath != "" {
+				file, err := os.OpenFile(outpath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, output_perm)
+				defer file.Close()
```
The `defer` call should go after checking the error. Otherwise it may call `Close()` on `nil` object. Same for the `COMMAND_QUERY_MEASUREMENT` case.

vasiliy-ul commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,280 @@
+				file, err := os.OpenFile(outpath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, output_perm)
+				defer file.Close()
+				if err != nil {
+					log.Log.Errorf("[SEV] Failed to write certificate chain to %s. File already exists", outpath)
```
You may have other errors here apart from EXISTS. E.g. EPERM. So the log message may be confusing. I would suggest to write it in a more generic way: 
```go
log.Log.Errorf("[SEV] Failed to write certificate chain to %s: %v", outpath, err)
```
Same for the `COMMAND_QUERY_MEASUREMENT` case.

Also a though about using `log.Log.Errorf(...)` in general. Does it make sense to log the error twice? I think by doing `return fmt.Errorf("Error %s/%s: %v", namespace, vmiName, err)` it will be logged automatically.

fabiand commented:
@alicefr thoughts?

vladikr commented:
@jean-edouard @iholder-redhat

iholder-redhat commented:
> @jean-edouard @iholder-redhat

Thanks for the FYI!
/assign @iholder-redhat

alicefr commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,290 @@
+		}
+		defer file.Close()
+
+		if sevPlatformInfo, err := virtClient.VirtualMachineInstance(namespace).SEVFetchCertChain(vmiName); err != nil {
```
I'd rewrite this as:
```golang
sevPlatformInfo, err := virtClient.VirtualMachineInstance(namespace).SEVFetchCertChain(vmiName)
if err != nil {
    return fmt.Errorf("Error %s/%s: %v", namespace, vmiName, err)
}
```
It can avoid one level indentation

alicefr commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,290 @@
+
+	case COMMAND_FETCH_CERT_CHAIN:
+		var file *os.File
+		pathProvided := outpath != ""
```
Can you evaluate this statement before the switch as you are repeating it 3 times there?

alicefr commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,290 @@
+				return fmt.Errorf("Error %s/%s: %v", namespace, vmiName, err)
+			}
+		}
+		defer file.Close()
```
Should be this placed inside the `if pathProvided` as otherwise you don't open the file?Just wondering if you don't get a nil pointer?

alicefr replied:
Genenrally for the write and open of the file you could create a function since you are repeating it. It could also help improving the readbility of he code

alicefr commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -15,143 +16,127 @@ import (
-			workDir = ""
-		}
+		vmi = api.NewMinimalVMI(vmiName)
+		workDir = GinkgoT().TempDir()
```
oh I see here the GinkgoT().TempDir :) , maybe could you rebase the fix in the original commit? It would make it easier for reviewers

alicefr commented:
@Fuzzy-Math I left a couple of comment. Generally, could you rebase the fix you did in the last 3 commits in the original ones? 
Additionally, do you plan to add any docs? I think it could be useful for the review if you can write some docs how to use the new commands with some examples

Fuzzy-Math commented:
@alicefr, I have addressed the majority of your comments, although some of them appeared to refer to older commits in the patch series. This was my bad for trying to keep clean conversation history over clean patches. I have rebased everything cleanly so hopefully it is a lot easier to review. 

In terms of docs: documenting SEV functionality is in the pipeline but out of scope for this PR. However, for the sake of this PR, I am maintaining a [cherry-picked branch](https://github.com/Fuzzy-Math/kubevirt/tree/sev-base) in order to build and run these changes. Then utilize `virtctl sev --help` to get a good summary of the commands and options that are available.

alicefr commented:
Thanks for the rebase, it looks good to me!

> In terms of docs: documenting SEV functionality is in the pipeline but out of scope for this PR. However, for the sake of this PR, I am maintaining a [cherry-picked branch](https://github.com/Fuzzy-Math/kubevirt/tree/sev-base) in order to build and run these changes. Then utilize `virtctl sev --help` to get a good summary of the commands and options that are available.

You are adding 4 new commands and this will be the most visible part to the users. So, I still think it will be useful to add the flow how those commands should be used and in which sequence with examples. But it could be part as a later PR

iholder-redhat commented:
/remove-lifecycle rotten

Still waiting for AMD hardware to be part of our CI.

xpivarc commented:
/reopen
/remove-lifecycle rotten

dhiller commented:
/reopen
/remove-lifecycle rotten

/test pull-kubevirt-e2e-k8s-1.26-sev

@Fuzzy-Math @alicefr @iholder101 @xpivarc IIUC we now have a manual test lane for AMD SEV testing in place.

Up to you folks whether you want to continue pushing this one.

vasiliy-ul commented:
> /reopen /remove-lifecycle rotten
> 
> /test pull-kubevirt-e2e-k8s-1.26-sev
> 
> @Fuzzy-Math @alicefr @iholder101 @xpivarc IIUC we now have a manual test lane for AMD SEV testing in place.
> 
> Up to you folks whether you want to continue pushing this one.

https://github.com/kubevirt/kubevirt/pull/7197 is now merged. I think this PR is now also good to go (after rebase and another review round).

alicefr commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,295 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022
```
This probably can be updated

alicefr commented on pkg/virtctl/sev/attestation.go:
```diff
@@ -0,0 +1,295 @@
+		return nil
+	}
+
+	if outpath != "" {
+		file, err := os.OpenFile(outpath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, output_perm)
+		if err != nil {
+			return fmt.Errorf(errFmt, namespace, vmiName, outpath, err)
+		}
+		defer file.Close()
+
+		_, err = file.Write(data)
+		if err != nil {
+			return fmt.Errorf(errFmt, namespace, vmiName, outpath, err)
+		}
+		log.Log.Infof(logFmt, outpath)
+	} else {
+		fmt.Printf("%s\n", string(data))
+	}
```
```suggestion
        if output == "" {
                fmt.Printf("%s\n", string(data))
                return nil
        }
        file, err := os.OpenFile(outpath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, output_perm)
        if err != nil {
                return fmt.Errorf(errFmt, namespace, vmiName, outpath, err)
        }
        defer file.Close()

        _, err = file.Write(data)
        if err != nil {
                return fmt.Errorf(errFmt, namespace, vmiName, outpath, err)
        }
        log.Log.Infof(logFmt, outpath)
```

alicefr commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,318 @@
+
+	Describe("The 'setup-session' command", func() {
+		Context("Creation", func() {
+			It("should succeed", func() {
```
Do we need this test?

alicefr commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,318 @@
+				Expect(cmd).ToNot(BeNil())
+			})
+		})
+		Context("With no argument", func() {
```
What about grouping this and the next 3 tests into a table?

alicefr commented on pkg/virtctl/sev/attestation_test.go:
```diff
@@ -0,0 +1,318 @@
+			})
+		})
+		Context("With no argument", func() {
+			It("should fail", func() {
```
Again, as above, try to group test that has very similar code into a table

alicefr commented:
@vasiliy-ul @Fuzzy-Math do we plan to add also functional tests for this?

alicefr commented:
/ok-to-test

vasiliy-ul commented:
> @vasiliy-ul @Fuzzy-Math do we plan to add also functional tests for this?

I think yes, now it makes sense since the API is in place.

dhiller commented:
@Fuzzy-Math I think a run of `make generate` and committing the changes is missing here: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7595/pull-kubevirt-generate/1677238768936947712#1:build-log.txt%3A1396

dhiller commented:
There's CI support on a periodic base for this here: https://testgrid.k8s.io/kubevirt-periodics#periodic-kubevirt-e2e-k8s-1.26-sev

/reopen
/remove-lifecycle rotten

dhiller commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

Let's see how the presubmit goes

dhiller commented:
Hey @vasiliy-ul @Fuzzy-Math, are you folks up to picking this up again?

vasiliy-ul commented:
> Hey @vasiliy-ul @Fuzzy-Math, are you folks up to picking this up again?

Hi @dhiller. Thank you for the ping. The PR probably needs a rebase and also some updates to address the comments.  Yes, definitely, we will look at it.

dhiller commented:
@vasiliy-ul I've converted it back to draft, just to reduce spent CI resources. Reconvert any time when it's ready.
